### PR TITLE
RFC: Include python3.def itself in the Rust source

### PR DESCRIPTION
**Pull request by adamreichold**
All parsing and processing is done in the CI so that the crate itself includes the python3.def file itself instead of creating it on the fly.

Based on the idea voiced in https://github.com/PyO3/python3-dll-a/pull/9#issuecomment-1113238455

I did try to keep the existing semantics in place and just moved things around to a Python script running in the CI. I just chose Python because I assume that the `ubuntu-latest` image contains a reasonably up-to-date version. (The updated CI workflow is untested, but I used the script locally the produce the definition file committed here.)

One could try using the [upstream script](https://github.com/python/cpython/blob/main/Tools/scripts/stable_abi.py) to do the parsing and generating, but I think that should be evaluated separately.

### CODE REVIEW

**Comment by ravenexp** [comment]
I agree with this change in principle, but I'd like to see at least some test coverage to ensure that the number of Stable ABI exports is non-decreasing with each definitions update PR. The Rust version had something like that.

You should also update `LICENSE` to remove all mentions of Python sources.

**Comment by adamreichold** [comment]
> I agree with this change in principle, but I'd like to see at least some test coverage to ensure that the number of Stable ABI exports is non-decreasing with each definitions update PR. The Rust version had something like that.

Added this as an `assert` whenever the parser script is run.

> You should also update `LICENSE` to remove all mentions of Python sources.

Done.

**Comment by adamreichold** [comment]
Also note that the automatically created PR will contain a diff of the generated definitions file this way, so we can also review that before merging.

**Comment by encukou** [comment]
The switch to TOML will happen today or early next week.

**Review by ravenexp — COMMENTED** [review]
Ignore this.

File: LICENSE [hunk 1]
```diff
@@ -1,7 +1,6 @@
 MIT License
 
-Copyright (c) 2022 Sergey Kvachonok (except file 'Misc/stable_abi.txt', which is
-Copyright (c) 2022 Python Software Foundation, see 'Misc/LICENSE' for its license)
+Copyright (c) 2022 Sergey Kvachonok
```
Comment (ravenexp) [inline LICENSE]: You should probably add yourself here as well, because you and @messense together authored like 2/3 of the current source code.

**Review by ravenexp — APPROVED** [review]
Looks good to me.

**Comment by adamreichold** [comment]
> The switch to TOML will happen today or early next week.

Thanks for the heads-up! I think we could do `import toml` then without adding any dependencies to the Rust build scripts. (Or do something simpler, I just mean that we unrestricted w.r.t. our dependency closure.)

**Comment by ravenexp** [comment]
> The switch to TOML will happen today or early next week.

Thanks for updating us on this. I'll delay the merge until the update script migrates to the TOML source.

File: LICENSE (hunk 1 above)
Comment (adamreichold) [reply LICENSE]: Added myself.

**Comment by encukou** [comment]
CPython just switched to TOML.

If you stick with Python for parsing, beware that [`toml`](https://pypi.org/project/toml/) is unmaintained/outdated.
Consider [`tomli`](https://pypi.org/project/tomli/), a little, correct TOML parser, which is being added into the Python 3.11 standard library as [`tomllib`](https://docs.python.org/dev/library/tomllib.html). (Full disclosure: I helped)

If you happen to know another project that needs a `python3.def` (or could use it), please let me know. I could probably generate it and keep it updated in CPython itself to reduce duplication of effort.

**Comment by ravenexp** [comment]
> If you happen to know another project that needs a `python3.def` (or could use it), please let me know. I could probably generate it and keep it updated in CPython itself to reduce duplication of effort.

Does this mean that `python3.def` will be committed to the CPython project repository and we will be able to pull it in as is without generating it ourselves from the TOML definitions?

**Comment by encukou** [comment]
Yes. CPython has some infrastructure to generate files based on `stable_abi.toml`, one more shouldn't be too much of a burden (though other devs might disagree).
A possible issue is that it wouldn't be *tested* in CPython, so we'd rely on you to report any issues.

Python generates and uses [python3dll.c](https://github.com/python/cpython/blob/main/PC/python3dll.c) instead. (Hm, would you be able to use that by any chance? I know next to nothing about building on Windows...)
edit: The exact formatting is as unstable as `stable_abi.toml`, it's only better if you can feed it to a C compiler directly.

**Comment by ravenexp** [comment]
> Python generates and uses [python3dll.c](https://github.com/python/cpython/blob/main/PC/python3dll.c) instead. (Hm, would you be able to use that by any chance? I know next to nothing about building on Windows...) edit: The exact formatting is as unstable as `stable_abi.toml`, it's only better if you can feed it to a C compiler directly.

No, I think it's specific to the Microsoft C++ compiler, and we also support MinGW and LLVM tools (without Clang). Also we only need the import library, not the DLL file itself.

**Review by ravenexp — APPROVED** [review]
Nice, thanks!

**Comment by ravenexp** [comment]
I think we can merge this now (?). I'll wait for @messense to take a look.

**Comment by adamreichold** [comment]
Had to switch to installing `python3-pip` via `apt-get` and `tomli` via pip as `ubuntu-latest` is 20.04 which does not yet contain the `python3-tomli` package introduced in 22.04.

**Comment by ravenexp** [comment]
> Had to switch to installing `python3-pip` via `apt-get` and `tomli` via pip as `ubuntu-latest` is 20.04 which does not yet contain the `python3-tomli` package introduced in 22.04.

TBH I was surprised that Ubuntu had up to date Python packages, but I'm mostly using official Rust docker images based on Debian Buster, so `pip install` was my go-to choice.

**Review by messense — APPROVED** [review]
👍

**Pull request merged**